File: shacs_bot/providers/transcription.py
"""Groq를 사용한 음성 전사 제공자."""
import os
from pathlib import Path

import httpx
from loguru import logger


class GroqTranscriptionProvider:
    """
    Groq의 Whisper API를 사용하는 음성 전사 제공자입니다.

    Groq는 매우 빠른 전사 속도를 제공하며, 넉넉한 무료 요금제를 지원합니다.
    """

    def __init__(self, api_key: str | None = None):
        self.api_key: str = api_key or os.environ.get("GROQ_API_KEY")
        self.api_url: str = "https://api.groq.com/openai/v1/audio/transcriptions"

    async def transcribe(self, file_path: str | Path) -> str:
        """
        Groq를 사용하여 오디오 파일을 전사합니다.

        Args:
            file_path: 오디오 파일의 경로.

        Returns:
            전사된 텍스트.
        """
        if not self.api_key:
            logger.warning("음성 전사를 위한 Groq API 키가 설정되지 않았습니다.")
            return ""

        path: Path = Path(file_path)
        if not path.exists():
            logger.error("오디오 파일을 찾지 못했습니다: {}", file_path)
            return ""

        try:
            async with httpx.AsyncClient() as client:
                with open(path, "rb") as f:
                    files = {
                        "file": (path.name, f),
                        "model": (None, "whisper-large-v3"),
                    }
                    headers = {
                        "Authorization": f"Bearer {self.api_key}",
                    }

                    response = await client.post(
                        self.api_url,
                        headers=headers,
                        files=files,
                        timeout=60.0
                    )
                    response.raise_for_status()

                    data: dict = response.json()
                    return data.get("text", "")

        except Exception as e:
            logger.error("Groq transcription 에러: {}", e)
            return ""

File: crates/shacs-providers/tests/transcription.rs
use serde_json::json;
use shacs_providers::{
    build_audio_transcription_request, parse_transcription_response, resolve_transcription_api_url,
    AudioTranscriptionHttpResponse, AudioTranscriptionRequestParts, GroqTranscriptionClient,
    ProviderError, TranscriptionClient, TranscriptionRequest,
};
use std::collections::BTreeMap;
use std::error::Error;
use std::fs;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

#[test]
fn resolve_transcription_url_accepts_full_endpoint_or_base() -> Result<(), Box<dyn Error>> {
    let joined = resolve_transcription_api_url(
        Some("https://api.groq.com/openai/v1/"),
        None,
        "https://fallback.invalid/audio/transcriptions",
    );
    let full = resolve_transcription_api_url(
        Some("https://example.test/v1/audio/transcriptions"),
        None,
        "https://fallback.invalid/audio/transcriptions",
    );
    let env_fallback = resolve_transcription_api_url(
        None,
        Some("https://env.example/v1"),
        "https://fallback.invalid/audio/transcriptions",
    );
    if joined != "https://api.groq.com/openai/v1/audio/transcriptions"
        || full != "https://example.test/v1/audio/transcriptions"
        || env_fallback != "https://env.example/v1/audio/transcriptions"
    {
        return Err(format!(
            "transcription URL resolution drifted: joined={joined} full={full} env={env_fallback}"
        )
        .into());
    }
    Ok(())
}

#[test]
fn build_audio_transcription_request_uses_openai_compatible_multipart() -> Result<(), Box<dyn Error>>
{
    let parts = build_audio_transcription_request(
        "https://api.example/v1/audio/transcriptions",
        "sk-test",
        "voice.ogg",
        b"audio-bytes",
        "whisper-large-v3",
        Some("ko"),
        Some("names are Korean"),
    );
    let body = String::from_utf8_lossy(&parts.body);
    if parts.url != "https://api.example/v1/audio/transcriptions"
        || parts.headers.get("Authorization").map(String::as_str) != Some("Bearer sk-test")
        || !parts
            .content_type
            .starts_with("multipart/form-data; boundary=shacs-bot-")
        || !body.contains("name=\"file\"; filename=\"voice.ogg\"")
        || !body.contains("audio-bytes")
        || !body.contains("name=\"model\"\r\n\r\nwhisper-large-v3")
        || !body.contains("name=\"response_format\"\r\n\r\njson")
        || !body.contains("name=\"language\"\r\n\r\nko")
        || !body.contains("name=\"prompt\"\r\n\r\nnames are Korean")
    {
        return Err(format!("multipart request drifted: parts={parts:?} body={body:?}").into());
    }
    Ok(())
}

#[test]
fn groq_transcription_client_posts_file_and_returns_text() -> Result<(), Box<dyn Error>> {
    let audio_path = temp_audio_path()?;
    fs::write(&audio_path, b"audio")?;
    let captured = Arc::new(Mutex::new(Vec::<AudioTranscriptionRequestParts>::new()));
    let captured_transport = captured.clone();
    let client = GroqTranscriptionClient::new(
        "gsk-test",
        "https://api.groq.com/openai/v1/audio/transcriptions",
        Some("ko".to_owned()),
        move |request: AudioTranscriptionRequestParts| {
            captured_transport
                .lock()
                .map_err(|error| ProviderError::Api {
                    status: None,
                    message: error.to_string(),
                    retryable: false,
                    headers: BTreeMap::new(),
                    body: None,
                })?
                .push(request);
            Ok(AudioTranscriptionHttpResponse {
                status: 200,
                headers: BTreeMap::new(),
                body: json!({"text": "안녕하세요"}).to_string(),
            })
        },
    );

    let mut request = TranscriptionRequest::new(audio_path.clone());
    request.prompt = Some("short greeting".to_owned());
    let text = client.transcribe(request)?;
    let _ = fs::remove_file(&audio_path);
    let captured = captured.lock().map_err(|error| error.to_string())?;
    let parts = captured.first().ok_or("missing transcription request")?;
    let body = String::from_utf8_lossy(&parts.body);
    if text != "안녕하세요"
        || !body.contains("name=\"model\"\r\n\r\nwhisper-large-v3")
        || !body.contains("name=\"language\"\r\n\r\nko")
        || !body.contains("short greeting")
    {
        return Err(format!("transcription client drifted: text={text:?} parts={parts:?}").into());
    }
    Ok(())
}

#[test]
fn parse_transcription_error_preserves_retryability_and_message() -> Result<(), Box<dyn Error>> {
    let error = match parse_transcription_response(AudioTranscriptionHttpResponse {
        status: 429,
        headers: BTreeMap::from([("retry-after".to_owned(), "2".to_owned())]),
        body: json!({"error": {"message": "rate limited"}}).to_string(),
    }) {
        Ok(value) => return Err(format!("expected error, got {value:?}").into()),
        Err(error) => error,
    };
    match error {
        ProviderError::Api {
            status,
            message,
            retryable,
            headers,
            ..
        } if status == Some(429)
            && message == "rate limited"
            && retryable
            && headers.get("retry-after").map(String::as_str) == Some("2") => {}
        other => return Err(format!("unexpected transcription error: {other:?}").into()),
    }
    Ok(())
}

fn temp_audio_path() -> Result<PathBuf, Box<dyn Error>> {
    let nanos = SystemTime::now().duration_since(UNIX_EPOCH)?.as_nanos();
    Ok(std::env::temp_dir().join(format!("shacs-transcription-{nanos}.ogg")))
}
